### ISSUE TITLE
Cake downloads multiple copies of Roslyn binaries (or fails to cleanup downloaded Roslyn binaries)

### ISSUE THREAD

**Issue by iKarthik** [cake-build/cake#1295]
What You Are Seeing?
Every time I run Cake.exe, it seems to download Roslyn binaries into %APPDATA/local/Temp/Nuget

What is Expected?
I dont expect CAKE to download the binaries (or clean up old ones), if they have been previously downloaded. Because of the current behavior, we have multiple copies of Roslyn binaries on our CI server (3000 copies in some cases)

What version of Cake are you using?
Currently on 0.11. I have verified similar behaviour exists on 0.16.2 as well

Are you running on a 32 or 64 bit system?
64 Bit

What environment are you running on?  Windows? Linux? Mac?
Windows

Are you running on a CI Server?  If so, which one?
Yes. Go.

How Did You Get This To Happen? (Steps to Reproduce)
1. Run the attached ./build.ps1. Go to  %APPDATA/local/Temp/Nuget, you will notice a a folder nested inside would be roslyn dlls.  (If you are on 0.16.2 it would download it to %AppData%\Local\Temp\NuGetScratch)
1. When you run the script above it would create a Cake folder in tools directory (as expected). Delete this cake folder 
2. Run the script again, you will notice new folders in  %APPDATA/local/Temp/Nuget which contains roslyn dlls in addition to the ones already there

This is akin to how our CI server runs the script. Every time we build, we nuke our directories (including tools folder), perform a clean fetch and then call the cake script

I understand moving the tools directory outside of our CI's scope is an option. I feel that this is a workaround, I suggest Cake should clean up the downloaded directories once its finished.

Thanks
K

**Comment by iKarthik** [comment]
Attached are the files. Renamed them because github wont let me upload .ps1 and .cake files

[build.txt](https://github.com/cake-build/cake/files/537508/build.txt)
[cake.txt](https://github.com/cake-build/cake/files/537509/cake.txt)

**Comment by devlead** [comment]
@iKarthik interesting, currently can't repro, so will need to investigate, the temp directory should be deleted after it's been copied to Cake folder, could be that something is holding on(like antivirus or similar) to the folder and that the delete temp folder fails silently.

https://github.com/cake-build/cake/blob/08907d1a5d97b66f58c01ae82506280882dcfacc/src/Cake/Scripting/Roslyn/Stable/RoslynScriptSessionFactory.cs#L120
https://github.com/cake-build/cake/blob/08907d1a5d97b66f58c01ae82506280882dcfacc/src/Cake/Scripting/Roslyn/Nightly/RoslynNightlyScriptSessionFactory.cs#L143

**Comment by devlead** [comment]
@iKarthik haven't been able to repro this but could potentially something like a antivirus/malware that's locking the folder.

Could you run a build with diagnostic verbosity so we can see if it's an silent fail or not?

**Comment by drinkbird** [comment]
I was able to reproduce the issue. In a blank folder, I downloaded the bootstrapper, created a cake project with a single target which fails intentionally, and executed the build. Then I checked everything in git, including `tools/Cake/*`.

When I run the build in Jenkins, I can see that cake re-downloads Roslyn from nuget, ignoring the fact that these files already exist in `tools/Cake/`. Here is the log:

```
c:\workspace\demo-freestyle>powershell .\build.ps1 -Target IntentionalError -Verbosity diagnostic 
Preparing to run build script...
Running build script...
Module directory does not exist.
Analyzing build script...
Analyzing C:/workspace/demo-freestyle/build.cake...
Processing build script...
Creating script session...
Downloading and installing Roslyn...
Installing packages (using https://packages.nuget.org/api/v2)...
Copying files...
Copying Roslyn.Compilers.CSharp.dll...
Copying Roslyn.Compilers.dll...
Deleting installation directory...
Adding reference to mscorlib.dll...
Adding reference to System.Core.dll...
Adding reference to Cake.Core.dll...
Adding reference to Cake.Common.dll...
Adding reference to Cake.exe...
Adding reference to System.dll...
Adding reference to System.Xml.dll...
Adding reference to System.Xml.Linq.dll...
Adding reference to System.Data.dll...
Importing namespace Cake.Common...
Importing namespace Cake.Common.Build...
Importing namespace Cake.Common.Build.AppVeyor...
Importing namespace Cake.Common.Build.AppVeyor.Data...
Importing namespace Cake.Common.Build.BitbucketPipelines...
Importing namespace Cake.Common.Build.BitbucketPipelines.Data...
Importing namespace Cake.Common.Build.Bitrise...
Importing namespace Cake.Common.Build.Bitrise.Data...
Importing namespace Cake.Common.Build.ContinuaCI...
Importing namespace Cake.Common.Build.ContinuaCI.Data...
Importing namespace Cake.Common.Build.GitLabCI...
Importing namespace Cake.Common.Build.GitLabCI.Data...
Importing namespace Cake.Common.Build.GoCD...
Importing namespace Cake.Common.Build.GoCD.Data...
Importing namespace Cake.Common.Build.Jenkins...
Importing namespace Cake.Common.Build.Jenkins.Data...
Importing namespace Cake.Common.Build.TFBuild...
Importing namespace Cake.Common.Build.TFBuild.Data...
Importing namespace Cake.Common.Build.TravisCI...
Importing namespace Cake.Common.Build.TravisCI.Data...
Importing namespace Cake.Common.Diagnostics...
Importing namespace Cake.Common.IO...
Importing namespace Cake.Common.IO.Paths...
Importing namespace Cake.Common.Net...
Importing namespace Cake.Common.Security...
Importing namespace Cake.Common.Solution...
Importing namespace Cake.Common.Solution.Project...
Importing namespace Cake.Common.Solution.Project.Properties...
Importing namespace Cake.Common.Solution.Project.XmlDoc...
Importing namespace Cake.Common.Text...
Importing namespace Cake.Common.Tools...
Importing namespace Cake.Common.Tools.Cake...
Importing namespace Cake.Common.Tools.Chocolatey...
Importing namespace Cake.Common.Tools.Chocolatey.ApiKey...
Importing namespace Cake.Common.Tools.Chocolatey.Config...
Importing namespace Cake.Common.Tools.Chocolatey.Features...
Importing namespace Cake.Common.Tools.Chocolatey.Install...
Importing namespace Cake.Common.Tools.Chocolatey.Pack...
Importing namespace Cake.Common.Tools.Chocolatey.Pin...
Importing namespace Cake.Common.Tools.Chocolatey.Push...
Importing namespace Cake.Common.Tools.Chocolatey.Sources...
Importing namespace Cake.Common.Tools.Chocolatey.Upgrade...
Importing namespace Cake.Common.Tools.DNU...
Importing namespace Cake.Common.Tools.DNU.Build...
Importing namespace Cake.Common.Tools.DNU.Pack...
Importing namespace Cake.Common.Tools.DNU.Restore...
Importing namespace Cake.Common.Tools.DotCover...
Importing namespace Cake.Common.Tools.DotCover.Analyse...
Importing namespace Cake.Common.Tools.DotCover.Cover...
Importing namespace Cake.Common.Tools.DotCover.Report...
Importing namespace Cake.Common.Tools.DotNetCore...
Importing namespace Cake.Common.Tools.DotNetCore.Build...
Importing namespace Cake.Common.Tools.DotNetCore.Execute...
Importing namespace Cake.Common.Tools.DotNetCore.Pack...
Importing namespace Cake.Common.Tools.DotNetCore.Publish...
Importing namespace Cake.Common.Tools.DotNetCore.Restore...
Importing namespace Cake.Common.Tools.DotNetCore.Run...
Importing namespace Cake.Common.Tools.DotNetCore.Test...
Importing namespace Cake.Common.Tools.DupFinder...
Importing namespace Cake.Common.Tools.Fixie...
Importing namespace Cake.Common.Tools.GitLink...
Importing namespace Cake.Common.Tools.GitReleaseManager...
Importing namespace Cake.Common.Tools.GitReleaseManager.AddAssets...
Importing namespace Cake.Common.Tools.GitReleaseManager.Close...
Importing namespace Cake.Common.Tools.GitReleaseManager.Create...
Importing namespace Cake.Common.Tools.GitReleaseManager.Export...
Importing namespace Cake.Common.Tools.GitReleaseManager.Publish...
Importing namespace Cake.Common.Tools.GitReleaseNotes...
Importing namespace Cake.Common.Tools.GitVersion...
Importing namespace Cake.Common.Tools.ILMerge...
Importing namespace Cake.Common.Tools.ILRepack...
Importing namespace Cake.Common.Tools.InnoSetup...
Importing namespace Cake.Common.Tools.InspectCode...
Importing namespace Cake.Common.Tools.MSBuild...
Importing namespace Cake.Common.Tools.MSTest...
Importing namespace Cake.Common.Tools.NSIS...
Importing namespace Cake.Common.Tools.NuGet...
Importing namespace Cake.Common.Tools.NuGet.Install...
Importing namespace Cake.Common.Tools.NuGet.Pack...
Importing namespace Cake.Common.Tools.NuGet.Push...
Importing namespace Cake.Common.Tools.NuGet.Restore...
Importing namespace Cake.Common.Tools.NuGet.SetApiKey...
Importing namespace Cake.Common.Tools.NuGet.SetProxy...
Importing namespace Cake.Common.Tools.NuGet.Sources...
Importing namespace Cake.Common.Tools.NuGet.Update...
Importing namespace Cake.Common.Tools.NUnit...
Importing namespace Cake.Common.Tools.OctopusDeploy...
Importing namespace Cake.Common.Tools.OpenCover...
Importing namespace Cake.Common.Tools.ReportGenerator...
Importing namespace Cake.Common.Tools.ReportUnit...
Importing namespace Cake.Common.Tools.Roundhouse...
Importing namespace Cake.Common.Tools.SignTool...
Importing namespace Cake.Common.Tools.SpecFlow...
Importing namespace Cake.Common.Tools.SpecFlow.StepDefinitionReport...
Importing namespace Cake.Common.Tools.SpecFlow.TestExecutionReport...
Importing namespace Cake.Common.Tools.TextTransform...
Importing namespace Cake.Common.Tools.VSTest...
Importing namespace Cake.Common.Tools.WiX...
Importing namespace Cake.Common.Tools.WiX.Heat...
Importing namespace Cake.Common.Tools.XBuild...
Importing namespace Cake.Common.Tools.XUnit...
Importing namespace Cake.Common.Xml...
Importing namespace Cake.Core...
Importing namespace Cake.Core.Diagnostics...
Importing namespace Cake.Core.IO...
Importing namespace Cake.Core.Scripting...
Importing namespace System...
Importing namespace System.Collections.Generic...
Importing namespace System.IO...
Importing namespace System.Linq...
Importing namespace System.Text...
Importing namespace System.Threading.Tasks...
Compiling build script...

========================================
IntentionalError
========================================
Executing task: IntentionalError
An error occurred when executing task 'IntentionalError'.
Error: System.Exception: An error to verify that cake exceptions will cause the build to fail as intended.
   at Submission#0.<.ctor>b__2()
   at Cake.Core.CakeTaskBuilderExtensions.<>c__DisplayClass5_0.<Does>b__0(ICakeContext context)
   at Cake.Core.ActionTask.Execute(ICakeContext context)
   at Cake.Core.DefaultExecutionStrategy.Execute(CakeTask task, ICakeContext context)
   at Cake.Core.CakeEngine.ExecuteTask(ICakeContext context, IExecutionStrategy strategy, Stopwatch stopWatch, CakeTask task, CakeReport report)
   at Cake.Core.CakeEngine.RunTarget(ICakeContext context, IExecutionStrategy strategy, String target)
   at Cake.Scripting.BuildScriptHost.RunTarget(String target)
   at Submission#0..ctor(Session session, Object& submissionResult)
   at Submission#0.<Factory>(Session session)
   at Roslyn.Scripting.CommonScriptEngine.Execute[T](String code, String path, DiagnosticBag diagnostics, Session session, Boolean isInteractive)
   at Roslyn.Scripting.Session.Execute(String code)
   at Cake.Scripting.Roslyn.Stable.DefaultRoslynScriptSession.Execute(Script script)
   at Cake.Core.Scripting.ScriptRunner.Run(IScriptHost host, FilePath scriptPath, IDictionary`2 arguments)
   at Cake.Commands.BuildCommand.Execute(CakeOptions options)
   at Cake.CakeApplication.Run(CakeOptions options)
   at Cake.Program.Main()

c:\workspace\demo-freestyle>exit 1 
Build step 'Execute Windows batch command' marked build as failure
Finished: FAILURE
```

**Comment by gep13** [comment]
@drinkbird is this on a public repo?  Can you confirm that the dll's and exe's did in fact make it into the repository?  Normally dll and exe files are ignored by default, so you would be no normally override those settings.  Can you confirm that this is what you did?

**Comment by drinkbird** [comment]
@gep13 It's a private repo but I could easily recreate the structure in a public one if necessary. Here's a screenshot of the Cake folder in jenkins, everything seems to be there. I also checked the git commit and the files have been added correctly.

![cake](https://cloud.githubusercontent.com/assets/1180736/22427726/279cc83e-e6fc-11e6-8586-9b2e59b4d61f.JPG)

Here's the interesting part. When I execute the build locally, it won't download Roslyn again. It only happens when the build is executed in Jenkins.
